### PR TITLE
feat: Collect logs during installation of Eclipse Che

### PR DESCRIPTION
**Pull request by tolusha**
### What does this PR do?
1. Introduce a new command: `server:logs` to collect Eclipse Che logs (except workspace logs)
2. Introduce  `workspace:logs` to collect workspace log
3. Automatically collect logs when Eclipse Che is being installed
4. Automatically collect logs when Eclipse Che Workspace is being started

```
logs
  |
  -- <namespace>
             |
              -------- <pod>
                            |
                            ----------- <container>.log
```

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/15485

### CODE REVIEW

File: src/commands/server/logs.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+export default class Logs extends Command {
+  static description = 'Retrieve Eclipse Che logs'
+
+  static flags = {
+    help: flags.help({ char: 'h' }),
+    chenamespace: cheNamespace,
+    'listr-renderer': listrRenderer,
+    'deployment-name': cheDeployment,
+    directory: string({
```
Comment (mmorhun) [inline src/commands/server/logs.ts]: Maybe it is better to be consistent and if we use quotes for some fields, let's use them in whole json?

File: src/tasks/che.ts [hunk 2]
```diff
@@ -464,4 +464,86 @@ export class CheTasks {
         }
       }]
   }
+
+  verifyCheNamespaceExistsTask(flags: any, command: Command): ReadonlyArray<Listr.ListrTask> {
+    return [{
+      title: `Verify if namespace ${flags.chenamespace} exists`,
+      task: async () => {
+        if (!await this.che.cheNamespaceExist(flags.chenamespace)) {
+          command.error(`E_BAD_NS - Namespace does not exist.\nThe Kubernetes Namespace "${flags.chenamespace}" doesn't exist. The configuration cannot be injected.\nFix with: verify the namespace where Che workspace is running (kubectl get --all-namespaces deployment | grep workspace)`, { code: 'EBADNS' })
```
Comment (mmorhun) [inline src/tasks/che.ts]: Maybe better to have underscore in error code, because I firstly read it as `Error BAd DNS`. Maybe me only, though...

Comment (mmorhun) [inline src/tasks/che.ts]: `E_BAD_NS - Namespace` - I think colon is more commonly used instead of dash, but up to you...

File: README.md [hunk 3]
```diff
@@ -185,6 +187,27 @@ OPTIONS
...
+  -h, --help  show CLI help
+
+  -n, --chenamespace=chenamespace
+      [default: che] Kubernetes namespace where Che server is supposed to be deployed
+
+  --deployment-name=deployment-name
+      [default: che] Che deployment name
+
+  -d, --directory=directory
+      [default: ./logs] Directory to store Eclipse Che logs
```
Comment (azatsarynnyy) [inline README.md]: Looks like readme isn't regenerated after the latest changes :slightly_smiling_face: 
see `yarn pack`

File: src/api/che.ts [hunk 4]
```diff
@@ -296,6 +297,96 @@ export class CheHelper {
     return ideURL.replace(/\/[^/|.]*\/[^/|.]*$/g, '\/dashboard\/#\/ide$&')
   }
 
+  /**
+   * Reads logs from all new pods that starting to run.
+   * It basically lists existed pods and starts following logs from a new ones.
+   */
+  async readAllNewPodLog(namespace: string, directory: string): Promise<void> {
+    const processedPods = new Set<string>((await this.kube.listNamespacedPod(namespace)).items.map(pod => pod.metadata!.name!))
+    setInterval(async () => this.readPodLogBySelectorIgnoreSpecificPods(namespace, undefined, directory, processedPods, true), 100)
```
[the remark this answers is not in the document]
Comment (tolusha) [reply src/api/che.ts]: 100 is just a polling interval and it works well.

File: src/api/che.ts [hunk 5]
```diff
@@ -296,6 +297,96 @@ export class CheHelper {
...
+    } else {
+      await this.readPodLogBySelectorIgnoreSpecificPods(namespace, selector, directory, processedPods, follow)
+    }
+  }
+
+  /**
+   * Reads logs from pods that match a given selector with exception of already processed ones.
+   * Once log is read the pod is marked as processed.
+   */
+  async readPodLogBySelectorIgnoreSpecificPods(namespace: string, selector: string | undefined, directory: string, processedPods: Set<string>, follow: boolean): Promise<void> {
```
[the remark this answers is not in the document]
Comment (tolusha) [reply src/api/che.ts]: `readPodLogBySelectorIgnoreProcessed`

File: src/api/kube.ts [hunk 6]
```diff
@@ -8,18 +8,20 @@
...
 
-import { ApiextensionsV1beta1Api, ApisApi, AppsV1Api, CoreV1Api, CustomObjectsApi, ExtensionsV1beta1Api, KubeConfig, RbacAuthorizationV1Api, V1beta1CustomResourceDefinition, V1beta1IngressList, V1ClusterRole, V1ClusterRoleBinding, V1ConfigMap, V1ConfigMapEnvSource, V1Container, V1DeleteOptions, V1Deployment, V1DeploymentList, V1DeploymentSpec, V1EnvFromSource, V1LabelSelector, V1ObjectMeta, V1PersistentVolumeClaimList, V1Pod, V1PodSpec, V1PodTemplateSpec, V1Role, V1RoleBinding, V1RoleRef, V1Secret, V1ServiceAccount, V1ServiceList, V1Subject } from '@kubernetes/client-node'
+import { ApiextensionsV1beta1Api, ApisApi, AppsV1Api, CoreV1Api, CustomObjectsApi, ExtensionsV1beta1Api, KubeConfig, Log, RbacAuthorizationV1Api, V1beta1CustomResourceDefinition, V1beta1IngressList, V1ClusterRole, V1ClusterRoleBinding, V1ConfigMap, V1ConfigMapEnvSource, V1Container, V1DeleteOptions, V1Deployment, V1DeploymentList, V1DeploymentSpec, V1EnvFromSource, V1LabelSelector, V1ObjectMeta, V1PersistentVolumeClaimList, V1Pod, V1PodList, V1PodSpec, V1PodTemplateSpec, V1Role, V1RoleBinding, V1RoleRef, V1Secret, V1ServiceAccount, V1ServiceList, V1Subject } from '@kubernetes/client-node'
```
[the remark this answers is not in the document]
Comment (tolusha) [reply src/api/kube.ts]: Unfortunately it is done automatically by `Organize Import` command

File: src/api/kube.ts [hunk 7]
```diff
@@ -424,13 +426,15 @@ export class KubeHelper {
...
     try {
-      const { body } = await k8sCoreApi.readNamespacedPod(name, namespace)
-      return this.compare(body, name)
+      const res = await k8sCoreApi.readNamespacedPod(name, namespace)
+      if (res && res.body) {
+        return res.body
+      }
     } catch {
-      return false
+      return
```
[the remark this answers is not in the document]
Comment (tolusha) [reply src/api/kube.ts]: Return value changed `boolean` -> `V1Pod | undefined`

File: src/commands/server/logs.ts (hunk 1 above)
Comment (tolusha) [reply src/commands/server/logs.ts]: Not possible due to warning
`Unnecessarily quoted property 'directory' found. (object-literal-key-quotes)tslint(1)`

File: src/tasks/che.ts (hunk 2 above)
Comment (tolusha) [reply src/tasks/che.ts]: I am not the author of this code. Just a copy/paste.

**Comment by tolusha** [comment]
@azatsarynnyy 
Could you approve the PR?

**Pull request merged**